/*
 * @begin:license
 *
 * Copyright (c) 2015-2018, Benjamin Niemann <pink@odahoda.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * @end:license
 */

syntax = "proto2";

import "noisicaa/audioproc/public/instrument_spec.proto";
import "noisicaa/audioproc/public/musical_time.proto";

package noisicaa.pb;

message ProcessorMessage {
  required string node_id = 1;

  message MuteNode {
    required bool muted = 1;
  }

  message PianoRollAddInterval {
    required uint64 id = 1;
    required MusicalTime start_time = 2;
    required MusicalTime end_time = 3;
    required uint32 pitch = 4;
    required uint32 velocity = 5;
  }

  message PianoRollRemoveInterval {
    required uint64 id = 1;
  }

  message CVGeneratorAddControlPoint {
    required uint64 id = 1;
    required MusicalTime time = 2;
    required float value = 3;
  }

  message CVGeneratorRemoveControlPoint {
    required uint64 id = 1;
  }

  message SampleScriptAddSample {
    required uint64 id = 1;
    required MusicalTime time = 2;
    required string sample_path = 3;
  }

  message SampleScriptRemoveSample {
    required uint64 id = 1;
  }

  message ChangeInstrument {
    required InstrumentSpec instrument_spec = 1;
  }

  oneof msg {
    MuteNode mute_node = 8;
    PianoRollAddInterval pianoroll_add_interval = 2;
    PianoRollRemoveInterval pianoroll_remove_interval = 3;
    CVGeneratorAddControlPoint cvgenerator_add_control_point = 4;
    CVGeneratorRemoveControlPoint cvgenerator_remove_control_point = 5;
    SampleScriptAddSample sample_script_add_sample = 6;
    SampleScriptRemoveSample sample_script_remove_sample = 7;
    ChangeInstrument change_instrument = 9;
  }
}

message ProcessorMessageList {
  repeated ProcessorMessage messages = 1;
}
